File: source/com/barrybecker4/simulation/trading/model/plugin/StrategyPlugins.java
/*
 * Copyright by Barry G. Becker, 2015. Licensed under MIT License: http://www.opensource.org/license.MIT
 */
package com.barrybecker4.simulation.trading.model.plugin;

import com.barrybecker4.common.util.PackageReflector;
import com.barrybecker4.simulation.trading.model.tradingstrategy.ITradingStrategy;

import java.io.IOException;
import java.lang.reflect.Modifier;
import java.security.AccessControlException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Finds plugins for type of stock series generation strategies.
 * Can be used to populate a dropdown.
 *
 * @author Barry Becker
 */
public class StrategyPlugins<E extends IStrategyPlugin> {


    private List<String> strategyNames = new ArrayList<>();
    private Map<String, E> valueMap = new HashMap<>();


    /**
     * Constructor
     * Uses reflection to dynamically load available strategies.
     *
     * @param packageName name of the package to get the plugin classes from
     *   Something like "com.barrybecker4.simulation.trading.model.generationstrategy"
     */
    public StrategyPlugins(String packageName, Class<?> clzz, List<E> defaultStrategies)  {

        try {
            List<Class> strategyClasses =
                    new PackageReflector().getClasses(packageName);

            for (Class<?> c : strategyClasses) {
                // Skip the abstract class (if any) because it cannot (and should not) be instantiated.
                if (!Modifier.isAbstract(c.getModifiers()) && clzz.isAssignableFrom(c)) {
                    E strategy = (E) c.newInstance();
                    String name = strategy.getName();
                    strategyNames.add(name);
                    valueMap.put(name, strategy);
                }
            }
            System.out.println("Trading strategy plugins found: " + valueMap);
        } catch (AccessControlException | NullPointerException e) {
            // fallback to local strategies if cannot access filesystem (as for applet or webstart)
            System.out.println("Unable to access filesystem to load plugin. Will use default strategies only");
            for (E s : defaultStrategies) {
                valueMap.put(s.getName(), s);
            }
        } catch (ClassNotFoundException | InstantiationException | IOException | IllegalAccessException e) {
            e.printStackTrace();
        }
    }


    public List<String> getStrategies() {
        return strategyNames;
    }

    /**
     * Create an instance of the algorithm given the controller and a refreshable.
     */
    public E getStrategy(String name) {
        if (!valueMap.containsKey(name)) {
            System.out.println("Could not find strategy with name " + name + " among " + valueMap);
        }
        return valueMap.get(name);
    }

}

File: scala-source/com/barrybecker4/simulation/trading/model/plugin/StrategyPlugins.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.trading.model.plugin

import com.barrybecker4.common.util.PackageReflector
import java.io.IOException
import java.lang.reflect.Modifier
import scala.collection.mutable.ListBuffer


/**
  * Finds plugins for type of stock series generation strategies.
  * Can be used to populate a dropdown.
  * Uses reflection to dynamically load available strategies.
  * @param packageName name of the package to get the plugin classes from
  *              Something like "com.barrybecker4.simulation.trading.model.generationstrategy"
  * @param clzz the type of class to look for. e.g. classOf[GenerationStrategy]
  * @param defaultStrategies list of default strategies to always include in the list.
  * @author Barry Becker
  */
class StrategyPlugins[E <: StrategyPlugin](val packageName: String, val clzz: Class[?],
                                           val defaultStrategies: Seq[E]) {
  private val strategyNames = ListBuffer[String]()
  private var valueMap = Map[String, E]()

  try {
    loadFromReflection()
  } catch {
    case _: NullPointerException | _: SecurityException =>
      useDefaultStrategiesOnly()
    case e@(_: ClassNotFoundException | _: InstantiationException | _: IOException | _: IllegalAccessException) =>
      e.printStackTrace()
  }

  mergeDefaultStrategies()

  private def registerStrategy(s: E): Unit = {
    if (!valueMap.contains(s.name)) {
      strategyNames.append(s.name)
    }
    valueMap += s.name -> s
  }

  private def loadFromReflection(): Unit = {
    val strategyClasses: Seq[Class[?]] = new PackageReflector().getClasses(packageName)
    for (c <- strategyClasses) {
      if (!Modifier.isAbstract(c.getModifiers) && clzz.isAssignableFrom(c)) {
        val strategy: E = c.getDeclaredConstructor().newInstance().asInstanceOf[E]
        registerStrategy(strategy)
      }
    }
  }

  private def useDefaultStrategiesOnly(): Unit = {
    valueMap = Map.empty
    strategyNames.clear()
    for (s <- defaultStrategies) {
      registerStrategy(s)
    }
  }

  /** Ensures built-in defaults appear in the map when reflection did not register them (e.g. alternate classloaders). */
  private def mergeDefaultStrategies(): Unit = {
    for (s <- defaultStrategies) {
      if (!valueMap.contains(s.name)) {
        registerStrategy(s)
      }
    }
  }

  def getStrategies: Seq[String] = strategyNames.toSeq

  /** Resolve a strategy by display name, or throw with a clear message. */
  def getStrategy(name: String): E =
    valueMap.getOrElse(
      name,
      throw new IllegalArgumentException(
        s"Unknown strategy '$name'. Known: ${valueMap.keys.toSeq.sorted.mkString(", ")}"
      )
    )
}
